### PR TITLE
ChibiOS support - part 5

### PR DESCRIPTION
**Pull request by tridge**
This PR is for continuing work on the ChibiOS port.

There are quite a few key features still needed:

[ ] chibios build of IOMCU firmware
[ ] px4pro support
[ ] fmuv1 support
[ ] io firmware upload
[ ] set_system_clock support
[ ] analog stop pins
[ ] serial5 support on fmuv3
[ ] i2c and ICU filtering (i2c filtering done)
[ ] PWM based rangefinders (in AP_RangerFinder_PWM_PX4)
[ ] nsh over mavlink for GPS, radio etc
[ ] safety switch and safety LED on boards without IOMCU

### CODE REVIEW

**Comment by Linjieqiang** [comment]
Maybe we need overclock option to make STM32F405 overclock frequency to 192M or more just like F4Light HAL Lib.

File: libraries/AP_HAL/Device.h [hunk 1]
```diff
@@ -78,6 +78,12 @@ class AP_HAL::Device {
         }
     }
 
+    /*
+     * Change device address. Note that this is the 7 bit address, it
+     * does not include the bit for read/write. Only works on I2C
+     */
+    virtual void set_address(uint8_t address) {};
```
Comment (lucasdemarchi) [inline libraries/AP_HAL/Device.h]: it doesn't really make sense to change a SPI address, does it? this should remain on the I2CDevice afaics

File: libraries/AP_Baro/AP_Baro_LPS2XH.cpp [hunk 2]
```diff
@@ -66,6 +73,62 @@ AP_Baro_Backend *AP_Baro_LPS2XH::probe(AP_Baro &baro,
...
+        return nullptr;
+    }
+    return sensor;
+}
+
+/*
+  setup invensense IMU to enable barometer, assuming both IMU and baro
+  on the same i2c bus
+*/
+bool AP_Baro_LPS2XH::_imu_i2c_init(uint8_t imu_address)
```
Comment (lucasdemarchi) [inline libraries/AP_Baro/AP_Baro_LPS2XH.cpp]: why does this need to be different than the other mag/baro on the aux bus?

File: libraries/AP_HAL/Device.h (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL/Device.h]: it is needed so we can have a generic driver that works with both bus types, but still allows set_address() on i2c.

Comment (lucasdemarchi) [reply libraries/AP_HAL/Device.h]: you have the bus type and you can act accordingly on where it is needed.

Comment (lucasdemarchi) [reply libraries/AP_HAL/Device.h]: on some drivers what I do is to have 2 constructors, one receiving I2CDevice and the other receiving SPIDevice. Then there's the common method that both call. Chances are high setting the address is not the only change you need to do anyway.  And in cases you don't want to differentiate the constructor you can still resort to
```c++
if (dev->bus_type == AP_HAL::Device::BUS_TYPE_I2C) { ... }
```

**Comment by davidbuzz** [comment]
all looks pretty reasonable to me.  :-)

**Pull request merged**